File: flut/.flutter/lib/flutter/widgets/scroll_physics.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/runtime.dart';
import 'package:flut/flut/object.dart';

class FlutScrollDecelerationRate
    extends FlutEnumObject<ScrollDecelerationRate> {
  const FlutScrollDecelerationRate()
    : super('ScrollDecelerationRate', const {
        'normal': ScrollDecelerationRate.normal,
        'fast': ScrollDecelerationRate.fast,
      });
}

class FlutScrollPhysics extends FlutValueObject {
  final ScrollPhysics value;
  const FlutScrollPhysics(this.value) : super('ScrollPhysics');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    if (value.parent != null) {
      result['parent'] = FlutScrollPhysics(value.parent!).flutEncode();
    }
    return result;
  }

  static ScrollPhysics? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return ScrollPhysics(
      parent: runtime.unpackOptionalField<ScrollPhysics>(data, 'parent'),
    );
  }
}

class FlutBouncingScrollPhysics extends FlutValueObject {
  final BouncingScrollPhysics value;
  const FlutBouncingScrollPhysics(this.value) : super('BouncingScrollPhysics');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    result['decelerationRate'] = const FlutScrollDecelerationRate().flutEncode(
      value.decelerationRate,
    )!;
    if (value.parent != null) {
      result['parent'] = FlutScrollPhysics(value.parent!).flutEncode();
    }
    return result;
  }

  static BouncingScrollPhysics? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return BouncingScrollPhysics(
      decelerationRate: runtime.unpackRequiredField<ScrollDecelerationRate>(
        data,
        'decelerationRate',
      ),
      parent: runtime.unpackOptionalField<ScrollPhysics>(data, 'parent'),
    );
  }
}

class FlutClampingScrollPhysics extends FlutValueObject {
  final ClampingScrollPhysics value;
  const FlutClampingScrollPhysics(this.value) : super('ClampingScrollPhysics');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    if (value.parent != null) {
      result['parent'] = FlutScrollPhysics(value.parent!).flutEncode();
    }
    return result;
  }

  static ClampingScrollPhysics? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return ClampingScrollPhysics(
      parent: runtime.unpackOptionalField<ScrollPhysics>(data, 'parent'),
    );
  }
}

class FlutNeverScrollableScrollPhysics extends FlutValueObject {
  final NeverScrollableScrollPhysics value;
  const FlutNeverScrollableScrollPhysics(this.value)
    : super('NeverScrollableScrollPhysics');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    if (value.parent != null) {
      result['parent'] = FlutScrollPhysics(value.parent!).flutEncode();
    }
    return result;
  }

  static NeverScrollableScrollPhysics? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return NeverScrollableScrollPhysics(
      parent: runtime.unpackOptionalField<ScrollPhysics>(data, 'parent'),
    );
  }
}

class FlutAlwaysScrollableScrollPhysics extends FlutValueObject {
  final AlwaysScrollableScrollPhysics value;
  const FlutAlwaysScrollableScrollPhysics(this.value)
    : super('AlwaysScrollableScrollPhysics');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    if (value.parent != null) {
      result['parent'] = FlutScrollPhysics(value.parent!).flutEncode();
    }
    return result;
  }

  static AlwaysScrollableScrollPhysics? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return AlwaysScrollableScrollPhysics(
      parent: runtime.unpackOptionalField<ScrollPhysics>(data, 'parent'),
    );
  }
}

File: flut/flutter/widgets/scroll_physics.py
from typing import Optional, override

from flut._flut.engine import FlutEnumObject, FlutValueObject, _flut_pack_value
from flut._flut.runtime import _flut_unpack_optional_field


class ScrollDecelerationRate(FlutEnumObject):
    normal: "ScrollDecelerationRate"
    fast: "ScrollDecelerationRate"


class ScrollPhysics(FlutValueObject):
    _flut_type = "ScrollPhysics"

    def __init__(
        self,
        *,
        parent: Optional["ScrollPhysics"] = None,
    ):
        super().__init__()
        self.parent = parent

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        if self.parent is not None:
            result["parent"] = _flut_pack_value(self.parent)
        return result

    @staticmethod
    def _flut_unpack(data: dict) -> "ScrollPhysics":
        return ScrollPhysics(
            parent=_flut_unpack_optional_field(data, "parent"),
        )


class BouncingScrollPhysics(ScrollPhysics):
    _flut_type = "BouncingScrollPhysics"

    def __init__(
        self,
        *,
        decelerationRate: ScrollDecelerationRate = ScrollDecelerationRate.normal,
        parent: Optional[ScrollPhysics] = None,
    ):
        super().__init__(parent=parent)
        self.decelerationRate = decelerationRate

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        result["decelerationRate"] = _flut_pack_value(self.decelerationRate)
        if self.parent is not None:
            result["parent"] = _flut_pack_value(self.parent)
        return result

    @staticmethod
    def _flut_unpack(data: dict) -> "BouncingScrollPhysics":
        return BouncingScrollPhysics(
            decelerationRate=_flut_unpack_optional_field(data, "decelerationRate")
            or ScrollDecelerationRate.normal,
            parent=_flut_unpack_optional_field(data, "parent"),
        )


class ClampingScrollPhysics(ScrollPhysics):
    _flut_type = "ClampingScrollPhysics"

    def __init__(
        self,
        *,
        parent: Optional[ScrollPhysics] = None,
    ):
        super().__init__(parent=parent)

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        if self.parent is not None:
            result["parent"] = _flut_pack_value(self.parent)
        return result

    @staticmethod
    def _flut_unpack(data: dict) -> "ClampingScrollPhysics":
        return ClampingScrollPhysics(
            parent=_flut_unpack_optional_field(data, "parent"),
        )


class NeverScrollableScrollPhysics(ScrollPhysics):
    _flut_type = "NeverScrollableScrollPhysics"

    def __init__(
        self,
        *,
        parent: Optional[ScrollPhysics] = None,
    ):
        super().__init__(parent=parent)

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        if self.parent is not None:
            result["parent"] = _flut_pack_value(self.parent)
        return result

    @staticmethod
    def _flut_unpack(data: dict) -> "NeverScrollableScrollPhysics":
        return NeverScrollableScrollPhysics(
            parent=_flut_unpack_optional_field(data, "parent"),
        )


class AlwaysScrollableScrollPhysics(ScrollPhysics):
    _flut_type = "AlwaysScrollableScrollPhysics"

    def __init__(
        self,
        *,
        parent: Optional[ScrollPhysics] = None,
    ):
        super().__init__(parent=parent)

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        if self.parent is not None:
            result["parent"] = _flut_pack_value(self.parent)
        return result

    @staticmethod
    def _flut_unpack(data: dict) -> "AlwaysScrollableScrollPhysics":
        return AlwaysScrollableScrollPhysics(
            parent=_flut_unpack_optional_field(data, "parent"),
        )
